src/protos/metadata.proto

```diff
@@ -19,6 +19,15 @@ message FileChunkMetadata {
   // An immutable and globally unique UUID chunk identifier.
   // This is assigned by the master at the time of chunk creation.
   string chunk_handle = 1;
+
+  // A version number associated with each chunk to distinguish between 
+  // up-to-date and stale replica
+  uint32 version = 2;
+
+  // The primary chunk server selected for a given chunk. The primary dicates
+  // mutation order among replicas and other replicas follo.
+  ChunkServerLocation primary_location = 3;
+  
   // All the replica locations for this particular file chunk
-  repeated ChunkServerLocation locations = 2;
+  repeated ChunkServerLocation locations = 4;
```

Comment:
Think chunk location should be handled by the chunk server manager. And the location can change, once the chunk server manager sees that a chunk server is down or when the chunk server doesn't report this chunk as part of it's chunks. I'm already adding code to get the locations for a given chunk handle. In the paper, it is also mentioned that chunk location isn't stored as part of metadata since the chunk server is the source of truth for the chunks it has. So I would suggest, once I introduce the api, you can get the locations of a chunk by just passing the chunk handle to the chunkserver manager. But you select which is primary. Thoughts?

Reply:
It sounds fine with me to move the chunk server location to chunk server manager. Just to clarify the possible coupling between MetadataManager and ChunkServerManager:

1) ChunkServerManager needs to hold a mapping from chunk handle (std::string) to ChunkServerLocation (maybe plus some other info). 

2) When MetadataManger creates a new chunk, it needs to create an entry in the ChunkServerLocation as an initialization. 

There is also some level of coupling for primary, I am fine with using MetadataManager to hold primary, but once it updates the primary, information also needs to be updated in ChunkServerLocation, for instance if this primary location is not in the list? 

If this sounds fine with you, I can purge out the location part. But you may want to take a snapshot of my code in case it can be reused for you (also the tests).

Reply:
Correct but let me clarify the APIs to avoid tight coupling between the two. The metadata manager will just be a user of the chunkservermgr. ChunkServerMgr will provide:
1. AllocateChunkServers(chunk_handle, number_of_servers): this will allocate the number of servers for the chunk_handle and returns the locations. Meaning servers where client can store the chunk. Will also store the new mapping from chunk_handle to the allocated chunk servers. Chunk will only get locations to store it when Allocate is called. Called when a new chunk is to be created. 
2. GetChunkLocations(chunk_handle): Get the locations where the chunk is stored. This can change from time to time, if a chunk_server has crashed.

ChunkServerMgr wouldn't store any metadata or primary chunk_server info. If MetadataMgr selected a primary previously and now called GetChunkLocations and didn't find the primary in the set of locations, it means the primary is down or corrupted and needs to select a new primary from the set returned.

And yes, chunkservermgr will  have a mapping from chunk_handle to locations as mentioned in 1. Does this work for you?

Reply:
Sounds good with me, I will remove the part that stores ChunkServerLocation from MetadataManager, and add placeholder code to call ChunkServerMgr API.